### DISCUSSION TITLE
Lazy loading JS trigeres

### DISCUSSION THREAD

**Discussion by khaledelmahdi** [livewire/livewire#6910] [1. Help]
I have a component the I lazy load on the page using the tag like this: `<livewire:component lazy />`

On the same page, I want to run a javascript code after the componenet finish loading (after the lazy place holder is removed).

This code fire directly after the component is initialized (which is correct):

```
<script>
    document.addEventListener('livewire:initialized', () => {
        console.log('here');
    });
</script>
```

Is there an event firered when the component finish lazy loading?

**Comment by WireNinja** [comment]
Hi @khaledelmahdi , you can use this alternative.

```php
<?php

#[Lazy]
class HeavyComponent extends Component
{
    public function mount()
    {
        $this->js(<<<JS
            console.log('test from lazy component')
            Livewire.dispatch('dispatched-from-lazy-component');
        JS); // work

        $this->dispatch('dispatch-for-blade'); // didnt work for me
        $this->dispatch('dispatch-for-blade')->self(); // didnt work for me as well
    }

    #[On('dispatched-from-lazy-component')]
    public function test()
    {
        $this->js(<<<JS
            console.log('console.log from lazy component');
        JS); // work
    }
}

// blade component
<script>
    Livewire.on('dispatch-for-blade', () => {
        console.log('dispatch for blade success'); // didnt work
    })
</script>
``` 

By utilizing mount, you can do anything when the component is fully loaded. But, in my journey so far, dispatching from livewire to blade component (in this case, nested component) won't make any event in the lazy loaded blade component. Juggling with javascript syntax inside livewire component is really challenging for me as well. 

Another way to taming complex javascript syntax by dispatching event to the parent. After listened to the event, do the js in the parent. 
Because of using '@this' inside javascript heredoc is hard as well. Last, Im sorry if my english is bad, hope this help.

**Comment by PhiloNL** [comment]
Try using Alpine instead.

If I put this inside my lazy loaded component it throws the alert once it's done loading:

```
<div x-data x-init="alert('lazy loaded')">
    I am lazy loaded
</div>
```

**Reply by WireNinja** [reply]
Wow, much more elegant and short. Thanks

**Reply by polycode-nz** [reply]
Is this the only solution for a nested components javascript dependancy?

It's not great.. This `console.log` example works but I have a hundred or so lines of JS I need in there.

**Reply by PhiloNL** [reply]
> Is this the only solution for a nested components javascript dependancy?
> 
> 
> 
> It's not great.. This `console.log` example works but I have a hundred or so lines of JS I need in there.
> 

Put your code in your app.js file and register an Alpine Data function.

Now you can simply do x-data="myDataFunction" and this will hook into your app.js code.

See the alpine docs and the data directive for more information.

**Reply by polycode-nz** [reply]
Isn't that in the face of the component based nature of livewire? I mean, if I do that - the components js code is there regardless if the component is on the page or not.

At the moment I'm writing js in a js view. Using CodeKit to compress it to a blade file and including the compressed js in the `x-init` of the components root div.

Strangely, if the code in x-init starts with `var` it errors with: `Unexpected token 'var'`.

If I put a redundant `console.log()` at the start, it works.

**Reply by kurowaffle** [reply]
Is there any alternative other than inserting the JS script in the x-init tag of Alpine? I have a lot of lines of javascript :(
Also, x-init or x-data just fired once. When component dyhrated, it'll not running again

**Comment by twmeier** [comment]
I have a similar issue. Lazy loading doesn't work as expected when needing to run scripts. 

For example, if I have this in my blade file:

```
<div>
    <div id="latest-orders-chart"></div>
</div>

@script
<script>
    console.log(document.getElementById('latest-orders-chart'));
</script>
@endscript
```
If I don't lazy load the component, the div exists and is logged in the console. If I do lazy load the component, I get "null" in the console. 

The script isn't waiting for the HTML to load before being triggered when lazy loaded.

**Reply by mxkh** [reply]
As a temporary workaround until the issue is resolved, try using setTimeout

```javascript
setTimeout(() => {
   console.log(document.querySelector('#your_id'));
}, 100);
```
This trick helped me to solve a problem, hope it'll work for others
